src/bls_ops.rs

```diff
@@ -0,0 +1,459 @@
+const BLS_PAIRING_COST_PER_ARG: Cost = 4515438;
+const BLS_MAP_TO_G1_BASE_COST: Cost = 610907;
+const BLS_MAP_TO_G1_COST_PER_BYTE: Cost = 122;
+const BLS_MAP_TO_G1_COST_PER_DST_BYTE: Cost = 135;
+const BLS_MAP_TO_G2_BASE_COST: Cost = 3380023;
+const BLS_MAP_TO_G2_COST_PER_BYTE: Cost = 122;
+const BLS_MAP_TO_G2_COST_PER_DST_BYTE: Cost = 135;
+
+const DST_G2: &[u8; 43] = b"BLS_SIG_BLS12381G2_XMD:SHA-256_SSWU_RO_AUG_";
+
+fn g1_atom(node: Node) -> Result<G1Affine, EvalErr> {
+    let blob = atom(node.clone(), "G1 atom")?;
+    if blob.len() != 48 {
+        return node.err("atom is not G1 size, 48 bytes");
+    }
+
+    match G1Affine::from_compressed(blob.try_into().expect("G1 slice is not 48 bytes")).into() {
+        Some(point) => Ok(point),
+        _ => node.err("atom is not a G1 point"),
+    }
+}
+
+fn g2_atom(node: Node) -> Result<G2Affine, EvalErr> {
+    let blob = atom(node.clone(), "G2 atom")?;
+    if blob.len() != 96 {
+        return node.err("atom is not G2 size, 96 bytes");
+    }
+
+    match G2Affine::from_compressed(blob.try_into().expect("G2 slice is not 96 bytes")).into() {
+        Some(point) => Ok(point),
+        _ => node.err("atom is not a G2 point"),
+    }
+}
+
+pub fn op_bls_g1_subtract(a: &mut Allocator, input: NodePtr, max_cost: Cost) -> Response {
+    let args = Node::new(a, input);
+    let mut cost = BLS_G1_SUBTRACT_BASE_COST;
+    check_cost(a, cost, max_cost)?;
+    let mut total: G1Projective = G1Projective::identity();
+    let mut is_first = true;
+    for arg in &args {
+        let point = g1_atom(arg)?;
+        cost += BLS_G1_SUBTRACT_COST_PER_ARG;
+        check_cost(a, cost, max_cost)?;
+        if is_first {
+            total = G1Projective::from(point);
+        } else {
+            total -= point;
+        };
+        is_first = false;
+    }
+    let total: G1Affine = total.into();
+    new_atom_and_cost(a, cost, &total.to_compressed())
+}
+
+pub fn op_bls_g1_multiply(a: &mut Allocator, input: NodePtr, max_cost: Cost) -> Response {
+    let args = Node::new(a, input);
+    check_arg_count(&args, 2, "g1_multiply")?;
+
+    let mut cost = BLS_G1_MULTIPLY_BASE_COST;
+    check_cost(a, cost, max_cost)?;
+
+    let mut total = G1Projective::from(g1_atom(args.first()?)?);
+    let args = args.rest()?;
+    let (scalar, scalar_len) = int_atom(args.first()?, "g1_multiply")?;
+    cost += scalar_len as Cost * BLS_G1_MULTIPLY_COST_PER_BYTE;
+    check_cost(a, cost, max_cost)?;
+
+    total *= number_to_scalar(mod_group_order(scalar));
+
+    let total: G1Affine = total.into();
+    new_atom_and_cost(a, cost, &total.to_compressed())
+}
+
+pub fn op_bls_g1_negate(a: &mut Allocator, input: NodePtr, max_cost: Cost) -> Response {
+    check_cost(a, BLS_G1_NEGATE_BASE_COST, max_cost)?;
+    let args = Node::new(a, input);
+    check_arg_count(&args, 1, "g1_negate")?;
+    let point = g1_atom(args.first()?)?;
+    let total = -point;
+    new_atom_and_cost(a, BLS_G1_NEGATE_BASE_COST, &total.to_compressed())
```

Comment:
@richardkiss is the compressed form guaranteed to just flip the most significant bit when negating? If so, perhaps it would make sense to implement this as just a validation check + the bit-flip. It's probably faster than `to_compressed()`.

Reply:
I believe so, yes. My understanding is that the encoding has a sign bit, so just figure out what that is and flip it. The only time this doesn't work is the zero (aka "infinity" just to be confusing) value, which I think has its own bit indicating 0, and (unsurprisingly) should pass through unchanged.